### PR TITLE
Fixes for index out of bounds exceptions.

### PR DESCRIPTION
**Pull request by ghost**
**Changes**
I made fixes to correct index out of bounds exceptions, added tests, and updated the changelog.

**Screenshots**
These screenshots show that the tests are passing.
Screen Shot 2020-01-16 at 12 55 15 PM

Screen Shot 2020-01-16 at 1 17 59 PM

### CODE REVIEW

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 1]
```diff
@@ -70,11 +71,13 @@ public void safeNotifyDataSetChanged() {
      * @param position Position of the item that has changed
      * @param payload Optional parameter, use null to identify a "full" update
      */
-    public void safeNotifyItemChanged(final int position, final Object payload) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemChangedWithPayloadRunnable(this, position, payload));
-        } else {
-            notifyItemChanged(position, payload);
+    public void safeNotifyItemChanged(final int position, @Nullable final Object payload) {
+        if (position >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 2]
```diff
@@ -84,10 +87,12 @@ public void safeNotifyItemChanged(final int position, final Object payload) {
      * @param position Position of the item that has changed
      */
     public void safeNotifyItemChanged(final int position) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemChangedRunnable(this, position));
-        } else {
-            notifyItemChanged(position);
+        if (position >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: safety check

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 3]
```diff
@@ -98,10 +103,12 @@ public void safeNotifyItemChanged(final int position) {
 
      */
     public void safeNotifyItemInserted(final int position) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemInsertedRunnable(this, position));
-        } else {
-            notifyItemInserted(position);
+        if (position >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: safety check

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 4]
```diff
@@ -112,10 +119,12 @@ public void safeNotifyItemInserted(final int position) {
      * @param toPosition New position of the item.
      */
     public void safeNotifyItemMoved(final int fromPosition, final int toPosition) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemMovedRunnable(this, fromPosition, toPosition));
-        } else {
-            notifyItemMoved(fromPosition, toPosition);
+        if (fromPosition >= 0 && toPosition >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 5]
```diff
@@ -127,10 +136,12 @@ public void safeNotifyItemMoved(final int fromPosition, final int toPosition) {
      * @param payload  Optional parameter, use null to identify a "full" update
      */
     public void safeNotifyItemRangeChanged(final int positionStart, final int itemCount, final Object payload) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemRangeChangedWithPayloadRunnable(this, positionStart, itemCount, payload));
-        } else {
-            notifyItemRangeChanged(positionStart, itemCount, payload);
+        if (positionStart >= 0 && itemCount >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 6]
```diff
@@ -141,10 +152,12 @@ public void safeNotifyItemRangeChanged(final int positionStart, final int itemCo
      * @param itemCount Number of items that have changed
      */
     public void safeNotifyItemRangeChanged(final int positionStart, final int itemCount) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemRangeChangedRunnable(this, positionStart, itemCount));
-        } else {
-            notifyItemRangeChanged(positionStart, itemCount);
+        if (positionStart >= 0 && itemCount >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 7]
```diff
@@ -155,10 +168,12 @@ public void safeNotifyItemRangeChanged(final int positionStart, final int itemCo
      * @param itemCount Number of items inserted
      */
     public void safeNotifyItemRangeInserted(final int positionStart, final int itemCount) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemRangeInsertedRunnable(this, positionStart, itemCount));
-        } else {
-            notifyItemRangeInserted(positionStart, itemCount);
+        if (positionStart >= 0 && itemCount >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 8]
```diff
@@ -169,10 +184,12 @@ public void safeNotifyItemRangeInserted(final int positionStart, final int itemC
      * @param itemCount Number of items removed from the data set
      */
     public void safeNotifyItemRangeRemoved(final int positionStart, final int itemCount) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemRangeRemovedRunnable(this, positionStart, itemCount));
-        } else {
-            notifyItemRangeRemoved(positionStart, itemCount);
+        if (positionStart >= 0 && itemCount >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 9]
```diff
@@ -182,10 +199,12 @@ public void safeNotifyItemRangeRemoved(final int positionStart, final int itemCo
      * @param position Position of the item that has now been removed
      */
     public void safeNotifyItemRemoved(final int position) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemRemovedRunnable(this, position));
-        } else {
-            notifyItemRemoved(position);
+        if (position >= 0) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: **safety check**

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/util/LinkedListUtil.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+     * @param <T>        the item type
+     * @return the item or null
+     */
+    public static <T> T getFirstItemQuietly(@Nullable LinkedList<T> linkedList) {
+        T item = null;
+
+        if (null != linkedList) {
+            try {
+                item = linkedList.getFirst();
+            } catch (NoSuchElementException e) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/util/LinkedListUtil.java]: **safety check**  Previously, this possibility was not being checked.

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickDataManager.java [hunk 11]
```diff
@@ -917,26 +1061,48 @@ public boolean hasInstanceOf(Class clazz) {
...
      * Checks / Determines if the brick is on the left wall, first row, right wall, last row.
      *
      * @param currentBrick BaseBrick item that was changed / added / removed
-     * @return index of first modified item
+     * @return index of first modified, non-null item or {@link #NO_PADDING_POSITION} for a null
+     * brick.
      */
-    private int computePaddingPosition(BaseBrick currentBrick) {
+    private int computePaddingPosition(@Nullable BaseBrick currentBrick) {
+        if (null == currentBrick) {
```
Comment (Greg-at-Wayfair) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickDataManager.java]: **safety check**  Previously, this possibility was not being checked.

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java (hunk 1 above)
Comment (kenyee) [reply BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: should this log a warning if position is < 0 to tell users they did something wrong?  Or if the library is broken?
Ditto the rest of these new safety checks?

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java [hunk 12]
```diff
@@ -70,11 +71,13 @@ public void safeNotifyDataSetChanged() {
...
      */
-    public void safeNotifyItemChanged(final int position, final Object payload) {
-        if (recyclerView.isComputingLayout()) {
-            handler.post(new NotifyItemChangedWithPayloadRunnable(this, position, payload));
-        } else {
-            notifyItemChanged(position, payload);
+    public void safeNotifyItemChanged(final int position, @Nullable final Object payload) {
+        if (position >= 0) {
+            if (recyclerView.isComputingLayout()) {
+                handler.post(new NotifyItemChangedWithPayloadRunnable(this, position, payload));
```
Comment (kenyee) [inline BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: sometimes I wonder if we should use DiffUtils to do all this automagically...

**Review by kenyee — APPROVED** [review]

File: BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java (hunk 1 above)
Comment (Greg-at-Wayfair) [reply BrickKit/bricks/src/main/java/com/wayfair/brickkit/BrickRecyclerAdapter.java]: Hey @kenyee I added in logging of warnings for the safety checks.

**Pull request merged**